daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -990,66 +999,128 @@ mkProjBindings env recExpr recTyp vsFlds e =
+        (aTo, aFrom) <- convertCoercion env coA
+        pure (fTo `ETmApp` aTo, fFrom `ETmApp` aFrom)
+    | otherwise = lift $ unhandled "Coercion" co
   where
-    go :: LF.Expr -> Coercion -> StateT Int ConvertM LF.Expr
-    go expr co
-      | isReflCo co
-      = pure expr
-
-      | Just (xCo, yCo) <- splitFunCo_maybe  co
-      = do
-        let Pair _ t = coercionKind xCo
-        n <- state (\k -> let l = k+1 in (l, l))
-        let var = mkVar ("$cast" ++ show n)
-        -- NOTE(MH): We need @mkSymCo xCo@ here because the arrow type is
-        -- contravariant in its first argument but GHC's coercions
-        -- don't reflect this fact.
-        var' <- go (EVar var) (mkSymCo xCo)
-        t' <- lift (convertType env t)
-        ETmLam (var, t') <$> go (expr `ETmApp` var') yCo
-
-      | Just (x, k_co, co') <- splitForAllCo_maybe co
-      , isReflCo k_co
-      = do
-        (x', kind) <- lift $ convTypeVar x
-        ETyLam (x', kind) <$> go (expr `ETyApp` TVar x') co'
-
-      -- Case (1) & (2)
-      | Pair s t <- coercionKind co
-      = lift $ do
-        mCase1 <- isSatNewTyCon s t
-        mCase2 <- isSatNewTyCon t s
-        case (mCase1, mCase2) of
-          (Just (tcon, field, flv), _) ->
-            let sanitize x
+    newtypeCoercion tCon ts field flv s0 t0 = do
+        sLf <- lift $ convertType env s0
+        tLf <- lift $ convertType env t0
+        ts' <- lift $ mapM (convertType env) ts
+        t' <- lift $ convertQualified env tCon
+        exprS <- mkLamBinder
+        exprT <- mkLamBinder
+        let tcon = TypeConApp t' ts'
+        let sanitizeTo x
                   -- NOTE(MH): This is DICTIONARY SANITIZATION step (3).
-                  | flv == ClassFlavour = ETmLam (mkVar "_", TUnit) x
-                  | otherwise = x
-            in pure $ ERecCon tcon [(field, sanitize expr)]
-          (_, Just (tcon, field, flv)) ->
-            let sanitize x
+                | flv == ClassFlavour = ETmLam (mkVar "_", TUnit) x
+                | otherwise = x
+            sanitizeFrom x
                   -- NOTE(MH): This is DICTIONARY SANITIZATION step (2).
-                  | flv == ClassFlavour = x `ETmApp` EUnit
-                  | otherwise = x
-            in pure $ sanitize (ERecProj tcon field expr)
-          _ -> unhandled "Coercion" co
-    isSatNewTyCon :: GHC.Type -> GHC.Type -> ConvertM (Maybe (TypeConApp, FieldName, TyConFlavour))
+                | flv == ClassFlavour = x `ETmApp` EUnit
+                | otherwise = x
+
+        pure $
+            ( ETmLam
+                  (exprS, sLf)
+                  (ERecCon tcon [(field, sanitizeTo (EVar exprS))])
+            , ETmLam
+                  (exprT, tLf)
+                  (sanitizeFrom (ERecProj tcon field (EVar exprT))))
+
+    Pair s t = coercionKind co
+
+    mkLamBinder = do
+        n <-
+            state
+                (\k ->
+                     let l = k + 1
+                      in (l, l))
```

Comment:
Can we please put this back onto a single line. :)

Reply:
`state (dupe . succ)`